File: backend/users/backtesting/indicators.py
"""
Technical indicators computed with pandas/numpy for backtesting.
"""

import numpy as np
import pandas as pd


def compute_rsi(series: pd.Series, period: int = 14) -> pd.Series:
    """Wilder's RSI."""
    delta = series.diff()
    gain = delta.where(delta > 0, 0.0)
    loss = (-delta).where(delta < 0, 0.0)
    avg_gain = gain.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    rs = avg_gain / avg_loss.replace(0, np.nan)
    return (100 - 100 / (1 + rs)).fillna(np.nan)


def compute_sma(series: pd.Series, period: int) -> pd.Series:
    return series.rolling(window=period).mean()


def compute_ema(series: pd.Series, period: int) -> pd.Series:
    return series.ewm(span=period, adjust=False).mean()


def compute_macd(series: pd.Series, fast: int = 12, slow: int = 26, signal: int = 9) -> dict[str, pd.Series]:
    ema_fast = compute_ema(series, fast)
    ema_slow = compute_ema(series, slow)
    macd_line = ema_fast - ema_slow
    signal_line = macd_line.ewm(span=signal, adjust=False).mean()
    histogram = macd_line - signal_line
    return {"MACD_LINE": macd_line, "MACD_SIGNAL": signal_line, "MACD_HIST": histogram}


def compute_bollinger(series: pd.Series, period: int = 20, std_dev: float = 2.0) -> dict[str, pd.Series]:
    middle = compute_sma(series, period)
    sd = series.rolling(window=period).std()
    upper = middle + std_dev * sd
    lower = middle - std_dev * sd
    bandwidth = ((upper - lower) / middle * 100).fillna(0)
    return {"BB_UPPER": upper, "BB_MIDDLE": middle, "BB_LOWER": lower, "BB_WIDTH": bandwidth}


def compute_atr(df: pd.DataFrame, period: int = 14) -> pd.Series:
    """Average True Range. Expects df with 'high', 'low', 'close' columns."""
    high_low = df["high"] - df["low"]
    high_close = (df["high"] - df["close"].shift(1)).abs()
    low_close = (df["low"] - df["close"].shift(1)).abs()
    tr = pd.concat([high_low, high_close, low_close], axis=1).max(axis=1)
    return tr.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()


def compute_stochastic(df: pd.DataFrame, k_period: int = 14, d_period: int = 3) -> dict[str, pd.Series]:
    """Stochastic %K and %D. Expects df with 'high', 'low', 'close' columns."""
    lowest_low = df["low"].rolling(window=k_period).min()
    highest_high = df["high"].rolling(window=k_period).max()
    k = ((df["close"] - lowest_low) / (highest_high - lowest_low).replace(0, np.nan)) * 100
    d = k.rolling(window=d_period).mean()
    return {"STOCH_K": k, "STOCH_D": d}


def compute_vwap(df: pd.DataFrame) -> pd.Series:
    """VWAP. Expects df with 'high', 'low', 'close', 'volume' columns."""
    tp = (df["high"] + df["low"] + df["close"]) / 3
    cum_vp = (tp * df["volume"]).cumsum()
    cum_vol = df["volume"].cumsum()
    return (cum_vp / cum_vol.replace(0, np.nan)).fillna(np.nan)


def compute_obv(df: pd.DataFrame) -> pd.Series:
    """On-Balance Volume. Expects df with 'close', 'volume' columns."""
    sign = np.sign(df["close"].diff())
    obv = (sign * df["volume"]).fillna(0).cumsum()
    return obv


def compute_indicator(config: dict, df: pd.DataFrame) -> dict[str, pd.Series]:
    """
    Compute a single indicator from its config dict.
    config: {"type": "RSI", "params": {"period": 14}, "source": "close"}
    df: DataFrame with OHLCV columns.
    Returns dict of key -> Series.
    """
    ind_type = config["type"]
    params = config.get("params", {})
    source_col = config.get("source", "close")
    source = df[source_col] if source_col in df.columns else df["close"]

    if ind_type == "RSI":
        key = f"RSI_{params.get('period', 14)}"
        return {key: compute_rsi(source, params.get("period", 14))}

    elif ind_type == "SMA":
        key = f"SMA_{params.get('period', 20)}"
        return {key: compute_sma(source, params.get("period", 20))}

    elif ind_type == "EMA":
        key = f"EMA_{params.get('period', 20)}"
        return {key: compute_ema(source, params.get("period", 20))}

    elif ind_type == "MACD":
        return compute_macd(source, params.get("fast", 12), params.get("slow", 26), params.get("signal", 9))

    elif ind_type == "BOLLINGER":
        return compute_bollinger(source, params.get("period", 20), params.get("stdDev", 2))

    elif ind_type == "ATR":
        key = f"ATR_{params.get('period', 14)}"
        return {key: compute_atr(df, params.get("period", 14))}

    elif ind_type == "STOCHASTIC":
        return compute_stochastic(df, params.get("kPeriod", 14), params.get("dPeriod", 3))

    elif ind_type == "VWAP":
        return {"VWAP": compute_vwap(df)}

    elif ind_type == "OBV":
        return {"OBV": compute_obv(df)}

    return {}

File: frontend-ui/src/lib/indicators.ts
/**
 * Pure-function technical indicators computed from OHLCV bars.
 *
 * Each function takes an array of bars (or a numeric series) and returns
 * a series of { time, value } points aligned with the input times.
 * Missing leading values (while warming up) are skipped, not zero-filled.
 */

export interface Bar {
  time: string | number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface IndicatorPoint {
  time: string | number;
  value: number;
}

export interface MultiIndicatorPoint {
  time: string | number;
  [key: string]: string | number;
}

// ---- Moving averages ----

export function sma(bars: Bar[], period: number, source: "close" | "open" | "high" | "low" = "close"): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length < period) return out;
  let sum = 0;
  for (let i = 0; i < bars.length; i++) {
    sum += bars[i][source];
    if (i >= period) sum -= bars[i - period][source];
    if (i >= period - 1) out.push({ time: bars[i].time, value: sum / period });
  }
  return out;
}

export function ema(bars: Bar[], period: number, source: "close" | "open" | "high" | "low" = "close"): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length < period) return out;
  const k = 2 / (period + 1);
  // Seed with SMA of first `period` values
  let prev = 0;
  for (let i = 0; i < period; i++) prev += bars[i][source];
  prev /= period;
  out.push({ time: bars[period - 1].time, value: prev });
  for (let i = period; i < bars.length; i++) {
    const next = bars[i][source] * k + prev * (1 - k);
    out.push({ time: bars[i].time, value: next });
    prev = next;
  }
  return out;
}

export function wma(bars: Bar[], period: number, source: "close" | "open" | "high" | "low" = "close"): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length < period) return out;
  const denom = (period * (period + 1)) / 2;
  for (let i = period - 1; i < bars.length; i++) {
    let weighted = 0;
    for (let j = 0; j < period; j++) {
      weighted += bars[i - j][source] * (period - j);
    }
    out.push({ time: bars[i].time, value: weighted / denom });
  }
  return out;
}

// ---- VWAP (session-based — resets each day) ----

export function vwap(bars: Bar[]): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  let cumPV = 0;
  let cumV = 0;
  let currentSession = "";
  for (const bar of bars) {
    const session = typeof bar.time === "string" ? bar.time.slice(0, 10) : new Date(Number(bar.time) * 1000).toISOString().slice(0, 10);
    if (session !== currentSession) {
      cumPV = 0;
      cumV = 0;
      currentSession = session;
    }
    const typical = (bar.high + bar.low + bar.close) / 3;
    cumPV += typical * bar.volume;
    cumV += bar.volume;
    if (cumV > 0) out.push({ time: bar.time, value: cumPV / cumV });
  }
  return out;
}

// ---- Bollinger Bands ----

export interface BollingerPoint {
  time: string | number;
  middle: number;
  upper: number;
  lower: number;
}

export function bollinger(bars: Bar[], period = 20, stdDev = 2): BollingerPoint[] {
  const out: BollingerPoint[] = [];
  if (bars.length < period) return out;
  for (let i = period - 1; i < bars.length; i++) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j++) sum += bars[j].close;
    const mean = sum / period;
    let variance = 0;
    for (let j = i - period + 1; j <= i; j++) {
      variance += (bars[j].close - mean) ** 2;
    }
    const sd = Math.sqrt(variance / period);
    out.push({
      time: bars[i].time,
      middle: mean,
      upper: mean + stdDev * sd,
      lower: mean - stdDev * sd,
    });
  }
  return out;
}

// ---- RSI ----

export function rsi(bars: Bar[], period = 14): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length <= period) return out;
  let gain = 0;
  let loss = 0;
  for (let i = 1; i <= period; i++) {
    const diff = bars[i].close - bars[i - 1].close;
    if (diff >= 0) gain += diff;
    else loss -= diff;
  }
  gain /= period;
  loss /= period;
  const firstRs = loss === 0 ? 100 : gain / loss;
  out.push({ time: bars[period].time, value: loss === 0 ? 100 : 100 - 100 / (1 + firstRs) });
  for (let i = period + 1; i < bars.length; i++) {
    const diff = bars[i].close - bars[i - 1].close;
    const g = diff > 0 ? diff : 0;
    const l = diff < 0 ? -diff : 0;
    gain = (gain * (period - 1) + g) / period;
    loss = (loss * (period - 1) + l) / period;
    const rs = loss === 0 ? 100 : gain / loss;
    out.push({ time: bars[i].time, value: loss === 0 ? 100 : 100 - 100 / (1 + rs) });
  }
  return out;
}

// ---- MACD ----

export interface MACDPoint {
  time: string | number;
  macd: number;
  signal: number;
  histogram: number;
}

export function macd(bars: Bar[], fast = 12, slow = 26, signalPeriod = 9): MACDPoint[] {
  const fastEma = ema(bars, fast);
  const slowEma = ema(bars, slow);
  if (slowEma.length === 0) return [];

  // Align fast EMA to slow EMA's start
  const offset = fastEma.length - slowEma.length;
  const macdLine: IndicatorPoint[] = [];
  for (let i = 0; i < slowEma.length; i++) {
    macdLine.push({
      time: slowEma[i].time,
      value: fastEma[i + offset].value - slowEma[i].value,
    });
  }

  // Signal = EMA of MACD line
  const signalLine: IndicatorPoint[] = [];
  if (macdLine.length >= signalPeriod) {
    const k = 2 / (signalPeriod + 1);
    let seed = 0;
    for (let i = 0; i < signalPeriod; i++) seed += macdLine[i].value;
    seed /= signalPeriod;
    signalLine.push({ time: macdLine[signalPeriod - 1].time, value: seed });
    let prev = seed;
    for (let i = signalPeriod; i < macdLine.length; i++) {
      const next = macdLine[i].value * k + prev * (1 - k);
      signalLine.push({ time: macdLine[i].time, value: next });
      prev = next;
    }
  }

  const signalOffset = macdLine.length - signalLine.length;
  const out: MACDPoint[] = [];
  for (let i = 0; i < signalLine.length; i++) {
    const macdVal = macdLine[i + signalOffset].value;
    const sig = signalLine[i].value;
    out.push({ time: signalLine[i].time, macd: macdVal, signal: sig, histogram: macdVal - sig });
  }
  return out;
}

// ---- Stochastic Oscillator ----

export interface StochasticPoint {
  time: string | number;
  k: number;
  d: number;
}

export function stochastic(bars: Bar[], kPeriod = 14, dPeriod = 3): StochasticPoint[] {
  const out: StochasticPoint[] = [];
  if (bars.length < kPeriod) return out;
  const kSeries: IndicatorPoint[] = [];
  for (let i = kPeriod - 1; i < bars.length; i++) {
    let high = -Infinity;
    let low = Infinity;
    for (let j = i - kPeriod + 1; j <= i; j++) {
      if (bars[j].high > high) high = bars[j].high;
      if (bars[j].low < low) low = bars[j].low;
    }
    const k = high === low ? 50 : ((bars[i].close - low) / (high - low)) * 100;
    kSeries.push({ time: bars[i].time, value: k });
  }
  for (let i = dPeriod - 1; i < kSeries.length; i++) {
    let sum = 0;
    for (let j = i - dPeriod + 1; j <= i; j++) sum += kSeries[j].value;
    out.push({ time: kSeries[i].time, k: kSeries[i].value, d: sum / dPeriod });
  }
  return out;
}

// ---- ATR (Average True Range) ----

export function atr(bars: Bar[], period = 14): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length <= period) return out;
  const trs: number[] = [];
  for (let i = 1; i < bars.length; i++) {
    const hl = bars[i].high - bars[i].low;
    const hc = Math.abs(bars[i].high - bars[i - 1].close);
    const lc = Math.abs(bars[i].low - bars[i - 1].close);
    trs.push(Math.max(hl, hc, lc));
  }
  let avg = 0;
  for (let i = 0; i < period; i++) avg += trs[i];
  avg /= period;
  out.push({ time: bars[period].time, value: avg });
  for (let i = period; i < trs.length; i++) {
    avg = (avg * (period - 1) + trs[i]) / period;
    out.push({ time: bars[i + 1].time, value: avg });
  }
  return out;
}

// ---- OBV (On-Balance Volume) ----

export function obv(bars: Bar[]): IndicatorPoint[] {
  const out: IndicatorPoint[] = [];
  if (bars.length === 0) return out;
  let total = 0;
  out.push({ time: bars[0].time, value: 0 });
  for (let i = 1; i < bars.length; i++) {
    if (bars[i].close > bars[i - 1].close) total += bars[i].volume;
    else if (bars[i].close < bars[i - 1].close) total -= bars[i].volume;
    out.push({ time: bars[i].time, value: total });
  }
  return out;
}

// ---- Heikin-Ashi transformation ----

export function toHeikinAshi(bars: Bar[]): Bar[] {
  const out: Bar[] = [];
  if (bars.length === 0) return out;
  // First bar stays
  out.push({ ...bars[0] });
  for (let i = 1; i < bars.length; i++) {
    const prev = out[i - 1];
    const cur = bars[i];
    const close = (cur.open + cur.high + cur.low + cur.close) / 4;
    const open = (prev.open + prev.close) / 2;
    const high = Math.max(cur.high, open, close);
    const low = Math.min(cur.low, open, close);
    out.push({ time: cur.time, open, high, low, close, volume: cur.volume });
  }
  return out;
}
